core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java

```diff
@@ -122,25 +123,20 @@ public boolean indexRoutingTableChanged(String index) {
      * Returns the indices deleted in this event
      */
     public List<Index> indicesDeleted() {
-        // If the new cluster state has a new cluster UUID, the likely scenario is that a node was elected
-        // master that has had its data directory wiped out, in which case we don't want to delete the indices and lose data;
-        // rather we want to import them as dangling indices instead.  So we check here if the cluster UUID differs from the previous
-        // cluster UUID, in which case, we don't want to delete indices that the master erroneously believes shouldn't exist.
-        // See test DiscoveryWithServiceDisruptionsIT.testIndicesDeleted()
-        // See discussion on https://github.com/elastic/elasticsearch/pull/9952 and
-        // https://github.com/elastic/elasticsearch/issues/11665
-        if (metaDataChanged() == false || isNewCluster()) {
+        if (metaDataChanged() == false) {
             return Collections.emptyList();
         }
         List<Index> deleted = null;
-        for (ObjectCursor<IndexMetaData> cursor : previousState.metaData().indices().values()) {
-            IndexMetaData index = cursor.value;
-            IndexMetaData current = state.metaData().index(index.getIndex());
-            if (current == null) {
+        final long previousVersion = previousState.version();
+        for (ObjectCursor<IndexTombstone> cursor : state.metaData().indexTombstones().values()) {
+            IndexTombstone tombstone = cursor.value;
+            // we should only try to delete indices that have tombstones added since
+            // the last time we processed cluster state
+            if (tombstone.getClusterVersion() > previousVersion) {
```

Comment:
I think we should compare the previous tombstone and the current one and generate a delta. Don't try to be overly smart.